### PR TITLE
Replace flake8 and pyupgrade with ruff

### PR DESCRIPTION
**Pull request by twoertwein**
Pandas switched to ruff a while ago: ruff is much faster and combines many different tools (including flake8-pyi and pyupgrade).

ruff fails currently because of this line which looks a bit weird:
https://github.com/pandas-dev/pandas-stubs/blob/89e083bf5548c6c944adbce74543437f7042c85c/pandas-stubs/io/formats/style.pyi#L43

### CODE REVIEW

File: scripts/test/run.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 
 
 def mypy_src():
-    cmd = ["mypy", "pandas-stubs", "tests", "--no-incremental"]
+    cmd = ["mypy", "pandas-stubs", "tests"]
```
Comment (twoertwein) [inline scripts/test/run.py:9]: It doesn't seem to fail anymore :) Running `poe mypy` twice takes for me 44.882s and then 0.710s.

**Review by Dr-Irv — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -10,29 +10,17 @@ repos:
...
+    -   id: ruff
         args: [
-          --ignore=E301 E302 E305 E402 E501 E701 E704 F401 F811 F821 W503 Y042,
-          # TypeVars in private files are already private
-          --per-file-ignores=_*.pyi:Y001
+          --exit-non-zero-on-fix,
+          --target-version, py38,
+          --extend-select, "PYI,UP,RUF100",
+          --ignore, "E501,E731,F841,PYI042",
+          --per-file-ignores, "_*.pyi:PYI001, _*.pyi:PYI030",
```
[the remark this answers is not in the document]
Comment (twoertwein) [reply .pre-commit-config.yaml:22]: I assume we want to keep them: https://github.com/pandas-dev/pandas-stubs/blob/89e083bf5548c6c944adbce74543437f7042c85c/pandas-stubs/_typing.pyi#L94

Will add line-based ignores instead.

[the remark this answers is not in the document]
Comment (Dr-Irv) [reply .pre-commit-config.yaml:22]: > Will add line-based ignores instead.

I agree that we should keep those, so using line-based ignores at least makes us think about it whenever they come up

File: pandas-stubs/core/indexes/accessors.pyi [hunk 3]
```diff
@@ -46,11 +46,14 @@ class Properties(PandasDelegate, PandasObject, NoNewAttributesMixin):
...
 class _DatetimeFieldOps(
-    _DayLikeFieldOps[_DTFieldOpsReturnType], _MiniSeconds[_DTFieldOpsReturnType]
+    _DayLikeFieldOps[_DTFieldOpsReturnType],  # noqa: F821
+    _MiniSeconds[_DTFieldOpsReturnType],  # noqa: F821
 ): ...
 class PeriodIndexFieldOps(
-    _DayLikeFieldOps[_IntIndexType],
-    _PeriodProperties[DatetimeIndex, _IntIndexType, Index, DatetimeIndex, PeriodIndex],
+    _DayLikeFieldOps[_IntIndexType],  # noqa: F821
+    _PeriodProperties[  # noqa: F821
```
[the remark this answers is not in the document]
Comment (twoertwein) [reply pandas-stubs/core/indexes/accessors.pyi:54]: Changed the order. This is a ruff (and flake8) bug - the order shouldn't matter inside a pyi file.

**Review by Dr-Irv — APPROVED** [review]
Thanks @twoertwein

**Comment by Dr-Irv** [comment]
@twoertwein  So I thanked you for this, then forgot to merge, and now there is a conflict. Can you resolve that conflict and push again?

**Review by Dr-Irv — APPROVED** [review]
thanks @twoertwein

**Pull request merged**